lib/auth/index.tsx

```diff
@@ -212,37 +214,127 @@ export class Auth extends Component<Props> {
+    // - after invalid, check at every input to see if it becomes valid
+    // see https://developer.mozilla.org/en-US/docs/Web/CSS/:-moz-ui-invalid
+
+    // monitor when the field has been changed by the user
+    if (event.currentTarget.value !== '') {
+      event.currentTarget.setAttribute('data-touched', true);
+    }
+
+    // we're using a className for validation so we can style in Electron
+    if (event.currentTarget.className !== 'validate') {
       return;
     }
+
+    const passwordError =
+      'Sorry, that password is not strong enough. Passwords must be at least 8 characters long and may not include your email address.';
+
+    if (this.state.isCreatingAccount) {
+      const username = get(this.usernameInput, 'value');
+      const password = get(this.passwordInput, 'value');
+
+      // run custom validation for the password but use the default message for missing value
+      if (
+        this.passwordInput.validity.valueMissing ||
+        validatePassword(password, username)
+      ) {
+        this.passwordInput.setCustomValidity('');
+      } else {
+        this.passwordInput.setCustomValidity(passwordError);
+      }
+    }
   };
 
-  onLogin = (event) => {
+  onSubmit = (event) => {
     event.preventDefault();
 
-    const username = get(this.usernameInput, 'value');
-    const password = get(this.passwordInput, 'value');
+    // clear any existing error messages on submit
+    this.props.resetErrors();
+    this.setState({
+      passwordErrorMessage: '',
+    });
 
-    if (!(username && password)) {
+    if (
+      this.usernameInput.validity.valueMissing ||
+      this.passwordInput.validity.valueMissing
+    ) {
       this.setState({
         passwordErrorMessage: 'Please fill out email and password.',
       });
       return;
     }
 
-    if (password.length < 4) {
+    // check that email is valid
+    if (!this.usernameInput.validity.valid) {
       this.setState({
-        passwordErrorMessage: 'Passwords must contain at least 4 characters.',
+        passwordErrorMessage: 'Please enter a valid email address.',
       });
       return;
     }
 
-    this.props.login(username, password);
+    const username = get(this.usernameInput, 'value');
+    const password = get(this.passwordInput, 'value');
+    const passwordError =
+      'Sorry, that password is not strong enough. Passwords must be at least 8 characters long and may not include your email address.';
+
+    // signup - stricter password requirements apply
+    if (this.state.isCreatingAccount) {
+      if (!validatePassword(password, username)) {
+        this.setState({
+          passwordErrorMessage: passwordError,
+        });
+        this.passwordInput.setCustomValidity(passwordError);
+        return;
+      }
+
+      this.passwordInput.setCustomValidity('');
+      this.props.signup(username, password, true);
+    }
+
+    // login - slightly more relaxed password rules
+    else {
+      if (!this.passwordInput.validity.valid) {
```

Comment:
why are we allowing relaxed passwords?

Reply:
We haven't implemented the check for forcing-reset-on-login yet... but even so, we first need to submit the auth, which means we need to allow existing passwords to go through

Reply:
Ahh, yes. I thought we do have the check: https://github.com/Automattic/simplenote-electron/pull/2088/

Reply:
Oh yeah I got confused about what was merged back into develop. Let's make sure to test this flow

Reply:
Yeah okay I still stand by this code. #2088 uses `authorize(username, password).then(...` which means the form needs to accept the insecure password in order to submit it for authorization, at which point we prompt them to reset. Confusing but I think we do need to do it, otherwise we don't know if it's the correct password to begin with.

Reply:
https://github.com/Automattic/simplenote-electron/pull/2088/files#diff-967127579e021461234f5de838baf700R47